### PR TITLE
New service tinia_winpak

### PR DESCRIPTION
**Pull request by stavamichal**
- generation script for tinia_winpak, generate data about firstName,
  lastName, login in MU namespace, information about access groups in
  tinia and gold account (if at least one resource has set it to true),
  chipNumber is primary key (unique) there
- send script for calling process script
- process script to execute imports, updates and deletes in target DB

### CODE REVIEW

File: gen/tinia_winpak [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/perl
+
+use strict;
+use warnings;
+use perunServicesInit;
+use perunServicesUtils;
+use Data::Dumper;
```
Comment (licehammer) [inline gen/tinia_winpak]: not required

File: gen/tinia_winpak [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		my %memberAttributes = attributesToHash $memberData->getAttributes;
+
+		my $firstName = $memberAttributes{$A_USER_FIRSTNAME};
+		my $lastName = $memberAttributes{$A_USER_LASTNAME};
+		my $loginMU = $memberAttributes{$A_USER_LOGIN_MU};
+		my $chipNumbers = $memberAttributes{$A_USER_CHIP_NUMBERS};
+
+		foreach my $chipNumber (@{$chipNumbers}) {
+			if(!defined($dataByChip->{$chipNumber})) {
+				$dataByChip->{$chipNumber}->{'FIRSTNAME'} = $firstName;
```
Comment (licehammer) [inline gen/tinia_winpak]: Please, use constant instead of hard-coded string 'FIRSTNAME'.

File: gen/tinia_winpak [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+				$dataByChip->{$chipNumber}->{'LASTNAME'} = $lastName;
+				if($resourceGold) { $dataByChip->{$chipNumber}->{'GOLD'} = 1; }
+				else { $dataByChip->{$chipNumber}->{'GOLD'} = 0; }
```
Comment (licehammer) [inline gen/tinia_winpak]: Can by simplified to:
````perl
$dataByChip->{$chipNumber}->{'GOLD'} = $resourceGold ? 1 : 0;
````

File: gen/tinia_winpak [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+				$dataByChip->{$chipNumber}->{'ACC_GROUPS'} = $resourceAccGroup;
+				$dataByChip->{$chipNumber}->{'LOGIN'} = $loginMU;
+			} else {
+				if($dataByChip->{$chipNumber}->{'LOGIN'} ne $loginMU) {
+					die "There are two same chips with different user logins in namespace MU: '" . $loginMU . "' and '" . $dataByChip->{$chipNumber}->{'LOGIN'} . "'!\n";
+				}
+
+				if($resourceGold) { $dataByChip->{$chipNumber}->{'GOLD'} = 1; }
+				$dataByChip->{$chipNumber}->{'ACC_GROUPS'} = $dataByChip->{$chipNumber}->{'ACC_GROUPS'} . "," . $resourceAccGroup;
```
Comment (licehammer) [inline gen/tinia_winpak]: This could lead to duplicates in ACC_GROUPS.

File: send/tinia_winpak_process.pl [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	print "Missing service file with data.\n";
+	exit 13;
+}
+
+my $configPath = "/etc/perun/services/$serviceName/$dbname";
+open FILE, $configPath or die "Could not open config file $configPath: $!";
+while(my $line = <FILE>) {
+	if($line =~ /^username: .*/) {
+		print "1\n";
+		$username = ($line =~ m/^username: (.*)$/x)[0];
```
Comment (licehammer) [inline send/tinia_winpak_process.pl]: If you use 'x' as regexp flag, then whitespaces in regexp are ignored, therefore matched username will include space in the beginning.

File: send/tinia_winpak_process.pl [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+my $allChips = join ',', @allChipsArray;
+
+my $chipsToDelete;
+if($allChips) {
+	$chipsToDelete = $dbh->prepare(qq{SELECT CardNumber from TINIA_WINPAK where AccessGroups IS NULL and CardType=1 and PROCESSED=0 and CardNumber NOT IN (?)});
+	$chipsToDelete->execute($allChips);
+} else {
+	#if service file is empty;
+	$chipsToDelete = $dbh->prepare(qq{SELECT CardNumber from TINIA_WINPAK where AccessGroups IS NULL and CardType=1 and PROCESSED=0});
+        $chipsToDelete->execute();
```
Comment (licehammer) [inline send/tinia_winpak_process.pl]: This line have wrong indentation.

**Comment by stavamichal** [comment]
Fixed.

**Pull request merged**